### ISSUE TITLE
Server Request Help: Instances vs References

### ISSUE THREAD

**Issue by Blazingfairy** [instaloader/instaloader#1624]
With Instaloader, one can create a Profile instance from a given username with the following snippet of code:

```
loader = Instaloader()
loader.load_session_from_file(login_name)
profile = Profile.from_username(loader.context, target_profile) 
```
Q1. Am I correct in assuming the last line makes a single request to the Instagram server? Am I also correct in assuming I am creating a data object (and not a reference) that is being stored in the "profile" variable?

`target_full_name = profile.full_name`

Q2. Am I correct in assuming this does not make a request to the server, but retrieves a parameter from the data object in the profile variable?

`followers = profile.get_followers()`

Q3. Am I correct in assuming this makes another request to the server and stores the data object in the "followers" variable? The documentation says the return type is "NodeIterator[Profile]". Does this mean I am storing the data object for each profile from the target's followers? Or am I making a reference to each of those profiles?

Q4. If I am storing the data object for each profile, does that mean extracting information from the data object for these profiles will NOT make additional requests to the server? (Because I already have it stored in the "followers" variable?) 
Or does it operate as a reference and thus: each moment I extract additional parameters from a follower profile, it will require an additional request to the server?

**Comment by aandergr** [comment]
> With Instaloader, one can create a Profile instance from a given username with the following snippet of code:
> 
> ```
> loader = Instaloader()
> loader.load_session_from_file(login_name)
> profile = Profile.from_username(loader.context, target_profile) 
> ```
> 
> Q1. Am I correct in assuming the last line makes a single request to the Instagram server? Am I also correct in assuming I am creating a data object (and not a reference) that is being stored in the "profile" variable?

Yes, `Profile.from_username()` does make a single request to Instagram, which contains most of the profile metadata, and returns a `Profile` instance with that data included.


> `target_full_name = profile.full_name`
> 
> Q2. Am I correct in assuming this does not make a request to the server, but retrieves a parameter from the data object in the profile variable?

Yes, `full_name` is already included in the data structure returned by `Profile.from_username()` and accessing it will in this case not trigger any further requests.

> `followers = profile.get_followers()`
> 
> Q3. Am I correct in assuming this makes another request to the server and stores the data object in the "followers" variable? The documentation says the return type is "NodeIterator[Profile]". Does this mean I am storing the data object for each profile from the target's followers? Or am I making a reference to each of those profiles?
>
> Q4. If I am storing the data object for each profile, does that mean extracting information from the data object for these profiles will NOT make additional requests to the server? (Because I already have it stored in the "followers" variable?)
> Or does it operate as a reference and thus: each moment I extract additional parameters from a follower profile, it will require an additional request to the server?

`get_followers()` makes one request to the server and returns a `NodeIterator`. Some of the followers are already included with a thin data structure, containing only basic information. Iterating this iterator or accessing the items it yields will trigger further requests as required.

**Comment by Blazingfairy** [comment]
Thank you so much! So:

```
   profile = Profile.from_username()
```
will already have the .userid, .username and .is_private properties, but will:
```            
followees = profile.get_followees()
```
also have those same properties? They seem like basic properties to me, but from reading the documentation and nodeiterator.py file I didn't discern that those properties are also retrieved (only the username, I think). So I am led to believe simply requesting the userid of each followee from followees would be an additional request - is that correct?
